### PR TITLE
feat(stereo): add native Fast Foundation Stereo

### PR DESCRIPTION
**Pull request by yifeif-nv**
## Background

Add isolated TensorRT Model Connect support for NVIDIA Fast Foundation Stereo, keep the complete production path on TensorRT native Network Definition APIs and model-owned plugins, and tune NVIDIA L4 performance without weakening accuracy gates.

## Exit Criteria

- Build and execute the complete feature, cost-volume, recurrent-update, and upsampling path without ONNX.
- Beat the recorded L4 baseline while retaining the published cosine, mean-error, and bad-2px thresholds.
- Record exhaustive whole-workload roofline evidence and fail closed when the strict 80/80 hardware-limit gate is not met.

## Implementation

- Builds the feature extractor, cost aggregation, geometry pyramid, eight recurrent updates, and convex upsampling directly with TensorRT Network Definition APIs.
- Uses no ONNX export, parser, intermediate artifact, or runtime path.
- Folds safe evaluation-time batch normalization and sinks decoder concatenation casts so the 88x88 and 176x176 boundaries remain FP16.
- Fuses normalized groupwise correlation plus left/right projected volumes in `FastFoundationStereoCombinedVolume`, retaining FP32 correlation accumulation.
- Adds `FastFoundationStereoGeometryVolumeConvc1` for the repeated geometry gather and first projection.
- Replaces the pathological channel reduction with `FastFoundationStereoSpatialAttentionReduce`, retaining FP32 mean/max accumulation.
- Adds `FastFoundationStereoPost8Sum` to combine and repack the full-resolution cost volume with bitwise-equivalent FP16 addition.
- Replaces three full-volume LeakyReLU pointwise launches with `FastFoundationStereoFullVolumeLeaky`, preserving FP32 alpha multiplication and FP16 rounding at verified DHWC8 boundaries.
- Lowers the disparity-head pointwise layers in NCHW, folds gamma into the final projection, and uses native `GELU_TANH` only for the measured second block.
- Selects the measured L4 builder configuration: feature optimization level 5 with two auxiliary streams; post optimization level 4 with no auxiliary streams; 8 GiB workspace.
- Keeps TensorRT 10's weakly typed post FP16 path with required FP32 tensors and a strongly typed fallback for releases without `BuilderFlag.FP16`.
- Keeps the accepted optimization paths on by default while retaining explicit environment-variable rollback switches.

## L4 Results

Hardware: NVIDIA L4, compute capability 8.9, 72 W limit, TensorRT 10.8.0.43, PyTorch 2.6.0+cu124.

Protocol: five fixed 700x700 stereo pairs, CUDA graphs, 20 warmup iterations and 100 timed iterations in each of five independent processes. Inference timing includes output conversion, device-to-host transfer, unpadding, and clipping.

| Metric | Result |
| --- | ---: |
| Recorded baseline, 5 pairs | 631.832829 ms |
| Native TensorRT, 500-sample mean | 253.935059 ms |
| Native TensorRT, 500-sample p90 | 255.513126 ms |
| Cross-process CV | 0.2305% |
| Throughput speedup | 2.488167x |
| Latency reduction | 59.8098% |
| Global cosine | 0.999798349 |
| Mean absolute disparity error | 0.336517 px |
| Bad-2px fraction | 0.0127261 |

Accuracy gates all pass: cosine >= 0.999, mean absolute error <= 0.5 px, and bad-2px fraction <= 0.02. All five independent runs produced the same output artifact hash.

## Roofline Evidence

Nsight Compute exhaustively captured all 774 eager feature-plus-post launches for one canonical pair with no launch filters, gaps, overlaps, or incomplete launches.

- Duration-weighted compute: 56.8696%
- Duration-weighted memory: 66.8440%
- Duration-weighted limiting ceiling: 73.1807%
- Duration at or above an 80% ceiling: 26.9644%
- Maximum individual-kernel compute/memory: 86.93% / 95.08%

The strict whole-workload 80/80 gate requires both weighted limiter >= 80% and at least 80% of duration near a ceiling. It fails. Some dominant kernels approach a hardware ceiling, but this PR does not claim the complete workload reached the L4 roofline.

## Validation

- `PYTHONPATH=python python3 -m pytest -q tests/e2e/models/fast_foundation_stereo --ignore=tests/e2e/models/fast_foundation_stereo/test_fast_foundation_stereo_e2e.py`: 198 passed.
- Receipt and Nsight analyzer regression suite: 31 passed.
- Rebase-added PixArt performance-matrix tests in the TensorRT container: 3 passed.
- Ruff check and format check: passed.
- CUDA/C++ clang-format dry run: passed.
- Final clean TensorRT plugin build and engine builds: passed on L4 with TensorRT 10.8 and CUDA 12.4.
- Five independent 100-iteration formal benchmark processes: passed accuracy and produced identical output hashes.
- Post8Sum full-volume oracle: random, structured, and signed-zero cases each compared 41,631,744 logical FP16 elements with zero bit mismatches.
- Full-volume LeakyReLU exhaustive half-domain oracle and three-site integration checks: bitwise equal; no added CopyNode at the plugin boundaries.
- Compute Sanitizer on the final custom-kernel set: memcheck and synccheck reported zero errors; racecheck reported zero hazards, errors, or warnings.
- Complete benchmark, engine, plugin, checkpoint, reference, input, profiler, and Nsight hashes are recorded in the schema-v2 L4 receipt.

## Notes For Future Readers

- Review the native graph boundaries first, then the four plugin registrations/implementations, then the model-owned tests and L4 receipt.
- The latency and accuracy milestone is qualified, but the strict roofline objective remains open; do not weaken the 80/80 gate or reinterpret the latency win as reaching the hardware limit.
- Current fork head: `8a9a3f8971d2988399a961fe80be9098784b1d91` from `yifeif-nv/TensorRT-Model-Connect:feat/fast-foundation-stereo`.
- The PR remains draft while exact-head GitHub CI runs. It must not be merged on the basis of older-head checks.

### CODE REVIEW

**Comment by yifeif-nv** [comment]
Superseded by #912. This draft's source repository detached from the NVIDIA fork network, so its PR head is frozen at `72bbe9c2bed524a513f5db77c2e930bf4a51610f` and cannot follow later pushes. The replacement draft is sourced from the real fork and currently points to exact head `8a9a3f8971d2988399a961fe80be9098784b1d91`. Please do not review, test, or merge this stale draft.
